### PR TITLE
cmd: `ollama config` command to help configure integrations to use Ollama

### PR DESCRIPTION
**Pull request by ParthSareen**
Adds a `ollama config` to allow easy an easy way to integrate (to start off) with:
- Droid
- Codex
- Claude
- Opencode

### CODE REVIEW

File: cmd/integrations.go [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+	Args: func(model string) []string {
+		if model == "" {
+			return nil
+		}
+		return []string{"--model", model}
+	},
+	CheckInstall: checkCommand("claude", "Install with: npm install -g @anthropic-ai/claude-code"),
+}
+
+var CodexConfig = &AppConfig{
```
Comment (drifkin) [inline cmd/integrations.go]: how about a version check on Codex and print a warning that they should update it if it's pre-Responses default for `--oss`? Otherwise it won't work well because of some bugs in Codex's deprecated chat completions support

File: cmd/apps.go [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+
+Supported apps:
+  claude    Claude Code
+  codex     Codex
+  droid     Droid
+  opencode  OpenCode
+
+Examples:
+  ollama connect
+  ollama connect claude
```
Comment (drifkin) [inline cmd/apps.go]: I didn't read the full PR yet, but when I read the claude part it looks like it's overriding env vars, which feels more like what I thought the `ollama launch` command was supposed to do. My thinking was `ollama connect` would durably configure an app to include some number of ollama models, whereas `ollama launch` would be ephemeral.

I think in the ephemeral case `ollama launch` giving the option to do some configuration on the first call would be a smoother flow (mainly: which model do you want to use?). In claude's case you could imagine wanting it to be durably configured to include ollama models without launching it via ollama, just the same as a GUI app would behave. Maybe my thinking is a little overcomplicated, but if we go for this current behavior I think we may want to rethink the terminology

File: cmd/selector.go [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+	"time"
+
+	"golang.org/x/term"
+
+	"github.com/ollama/ollama/api"
+)
+
+const maxDisplayedItems = 10
+
+var AppOrder = []string{"claude", "codex", "opencode", "droid"}
```
Comment (jmorganca) [inline cmd/selector.go]: Is this file for selector? but this is hardcoded here? I'd try and modularize this better

File: cmd/config.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type IntegrationConfig struct {
```
Comment (jmorganca) [inline cmd/config.go]: I think there's some renaming to do here (e.g. is the word integration still needed?)

File: cmd/integrations.go [hunk 5]
```diff
@@ -0,0 +1,785 @@
+package cmd
```
Comment (jmorganca) [inline cmd/integrations.go]: I'm not sure we should calls this apps.go vs just focusing on config + having methods for the different tools like droid

File: cmd/selector.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+	Description string
+}
+
+// terminalState manages raw terminal mode for interactive selection
+type terminalState struct {
+	fd       int
+	oldState *term.State
+}
+
+// enterRawMode puts terminal in raw mode with cursor hidden
```
Comment (jmorganca) [inline cmd/selector.go]: I think this exists in cmd elsewhere?

File: cmd/config/claude.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package integrations
```
Comment (jmorganca) [inline cmd/config/claude.go]: I'd put this under a config/ dir and package if that's not weird

File: cmd/config/codex.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
+package integrations
```
Comment (jmorganca) [inline cmd/config/codex.go]: I know these are ephemeral only but would be great to have tests.

File: cmd/config/selector.go [hunk 9]
```diff
@@ -0,0 +1,496 @@
+package integrations
```
Comment (jmorganca) [inline cmd/config/selector.go]: Test this on windows/cmd if you haven't yet

File: cmd/config/files.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
+package integrations
```
Comment (jmorganca) [inline cmd/config/files.go]: util.go is generally not a good name for a file. I'd keep this in integrations (or config.go as suggested)

File: cmd/config/files.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	backupPath := filepath.Join(backupDir, fmt.Sprintf("%s.%d", filepath.Base(srcPath), time.Now().Unix()))
+	if err := copyFile(srcPath, backupPath); err != nil {
+		return "", err
+	}
+	return backupPath, nil
+}
+
+func atomicWriteJSON(path string, data any) error {
```
Comment (jmorganca) [inline cmd/config/files.go]: I get the feeling you can break this up a bit and simplify it to not json but writing raw data and do the json marshal ahead of time

File: cmd/config/claude.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+			{Name: "ANTHROPIC_AUTH_TOKEN", Value: "ollama"},
+		}
+	},
+	Args: func(model string) []string {
+		if model == "" {
+			return nil
+		}
+		return []string{"--model", model}
+	},
+	CheckInstall: checkCommand("claude", "Install with: npm install -g @anthropic-ai/claude-code"),
```
Comment (BruceMacD) [inline cmd/config/claude.go]: The npm installation is deprecated, should use the native one now.

Should prob just link them here:
https://code.claude.com/docs/en/quickstart

File: cmd/config/claude.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package config
+
+var claudeIntegration = &integrationDef{
+	Name:        "Claude",
+	DisplayName: "Claude Code",
+	Command:     "claude",
+	EnvVars: func(model string) []envVar {
+		return []envVar{
+			{Name: "ANTHROPIC_BASE_URL", Value: "http://localhost:11434"},
+			{Name: "ANTHROPIC_API_KEY", Value: "ollama"},
```
Comment (BruceMacD) [inline cmd/config/claude.go]:
```
 ▐▛███▜▌   Claude Code v2.1.15
▝▜█████▛▘  minimax-m2.1:cloud · API Usage Billing
  ▘▘ ▝▝    ~/


 ⚠Auth conflict: Both a token (ANTHROPIC_AUTH_TOKEN) and an API key
  (ANTHROPIC_API_KEY) are set. This may lead to unexpected behavior.
    • Trying to use ANTHROPIC_AUTH_TOKEN? Unset the ANTHROPIC_API_KEY
    environment variable, or claude /logout then say "No" to the API
    key approval before login.
    • Trying to use ANTHROPIC_API_KEY? Unset the ANTHROPIC_AUTH_TOKEN
    environment variable.
```

Might be missing some kind of env config check here

Comment (ParthSareen) [reply cmd/config/claude.go]: I think I have a fix

File: cmd/config/config.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+type integrationConfig struct {
+	App          string    `json:"app"`
+	Models       []string  `json:"models"`
+	ConfiguredAt time.Time `json:"configured_at"`
+}
+
+// defaultModel returns the first (default) model, or empty string if none
+func (c *integrationConfig) defaultModel() string {
```
Comment (jmorganca) [inline cmd/config/config.go]: does this have to be a function? it returns the first one..

Comment (ParthSareen) [reply cmd/config/config.go]: Pulling this in a few different places so don't really want to do this bounds check everytime

File: cmd/config/codex.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+	"golang.org/x/mod/semver"
+)
+
+var codexIntegration = &integrationDef{
+	Name:        "Codex",
+	DisplayName: "Codex",
+	Command:     "codex",
+	EnvVars: func(model string) []envVar {
+		return []envVar{}
```
Comment (jmorganca) [inline cmd/config/codex.go]: `nil` should be default

File: cmd/config/droid.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+	Name:         "Droid",
+	DisplayName:  "Droid",
+	Command:      "droid",
+	EnvVars:      func(model string) []envVar { return nil },
+	Args:         func(model string) []string { return nil },
+	Setup:        setupDroidSettings,
+	CheckInstall: checkCommand("droid", "Install from: https://docs.factory.ai/cli/getting-started/quickstart"),
+}
+
+func isValidReasoningEffort(effort string) bool {
```
Comment (jmorganca) [inline cmd/config/droid.go]: Do we need this?

File: cmd/config/droid.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+var validReasoningEfforts = []string{"high", "medium", "low", "none"}
+
+func isValidReasoningEffort(effort string) bool {
+	return slices.Contains(validReasoningEfforts, effort)
+}
+
+func isOllamaModelEntry(m any) bool {
+	entry, ok := m.(map[string]any)
```
Comment (jmorganca) [inline cmd/config/droid.go]: Let's avoid casting here if we can.. I think everything we're working with is typed?

File: cmd/config/opencode.go [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+	Command:          "opencode",
+	EnvVars:          func(model string) []envVar { return nil },
+	Args:             func(model string) []string { return nil },
+	Setup:            setupOpenCodeSettings,
+	CheckInstall:     checkCommand("opencode", "install from https://opencode.ai"),
+	ConfigPaths:      openCodeConfigPaths,
+	ConfiguredModels: openCodeModels,
+}
+
+func setupOpenCodeSettings(modelList []string) error {
```
Comment (jmorganca) [inline cmd/config/opencode.go]: I wonder if just using 

```
func (o *opencode) Setup() {
... implementation
}
```

File: cmd/config/integrations.go [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+
+func handleCancelled(err error) (bool, error) {
+	if errors.Is(err, errCancelled) {
+		return true, nil
+	}
+	return false, err
+}
+
+// ConfigCmd returns the cobra command for configuring integrations.
+func ConfigCmd(checkServerHeartbeat func(cmd *cobra.Command, args []string) error) *cobra.Command {
```
Comment (jmorganca) [inline cmd/config/integrations.go:233]: I'd see if you can put this in top level cmd and have it call functions in this package.

File: cmd/config/integrations.go [hunk 20]
```diff
@@ -0,0 +1,383 @@
...
+		model, err := selectPrompt(fmt.Sprintf("Select model for %s:", integ.DisplayName), items)
+		if err != nil {
+			return nil, err
+		}
+		selected = []string{model}
+	}
+
+	for _, model := range selected {
+		if cloudModels[model] {
+			if err := ensureSignedIn(ctx, client); err != nil {
```
Comment (jmorganca) [inline cmd/config/integrations.go]: Can we get away without this check? E.g. only show cloud models if the user is signed in (do a check once?) this will check once per model selected

File: cmd/config/config.go [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+import (
+	"encoding/json"
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type integrationConfig struct {
```
Comment (jmorganca) [inline cmd/config/config.go]: It seems like this is only needed for non writable configs (e.g. claude code), as the droid/opencode configs have this list of models too

Given they launch with one model at a time, I don't think we need this., as much as a higher level config for ollama that has preferred model for each (or maybe we just default to the most recently model in ollama (might involve storing one name)?)

File: cmd/config/files.go [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	backupPath := filepath.Join(dir, fmt.Sprintf("%s.%d", filepath.Base(srcPath), time.Now().Unix()))
+	if err := copyFile(srcPath, backupPath); err != nil {
+		return "", err
+	}
+	return backupPath, nil
+}
+
+func atomicWrite(path string, data []byte) error {
```
Comment (jmorganca) [inline cmd/config/files.go]: I think a lot of this file can instead exist in the integrations. Each one might have specific ways to back up files or considerations. In the end we're writing a file and so the simpler that is the less error prone it may be - and we can have unit tests for each integration to test its backup capability (with some common functions like this called in config.go if needed)

File: cmd/config/integrations.go (hunk 20 above)
Comment (ParthSareen) [reply cmd/config/integrations.go]: I think we can do a check + save but IIUC we do have to ping at least once.

File: cmd/config/config.go (hunk 21 above)
Comment (ParthSareen) [reply cmd/config/config.go]: Not very tied to this but I think we should have them be the same as we probably will want to have a variety of models associated with an integration. TBD. Will see how it feels while I cleanup

File: cmd/config/files.go (hunk 22 above)
Comment (ParthSareen) [reply cmd/config/files.go]: I'd argue that having a read/write behavior per integration would be more complex. I think there's some tests + modifications which sit on top but the base behavior to ensure a file gets written correctly is the same. With that - I'll take another look to see if we can simplify

File: cmd/config/claude.go [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+	if _, err := exec.LookPath("claude"); err != nil {
+		return fmt.Errorf("claude is not installed, install from https://code.claude.com/docs/en/quickstart")
+	}
+
+	var args []string
+	if model != "" {
+		args = []string{"--model", model}
+	}
+
+	cmd := exec.Command("claude", args...)
```
Comment (jmorganca) [inline cmd/config/claude.go]: nit: does this let us use it like 

```
ollama config --launch claude make me a sandwhich
```?

OK if not, j/w

File: cmd/config/codex_test.go [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"testing"
+)
+
+func TestCodexIntegration(t *testing.T) {
+	c := &Codex{}
+
+	t.Run("String", func(t *testing.T) {
+		if got := c.String(); got != "Codex" {
```
Comment (jmorganca) [inline cmd/config/codex_test.go]: Not much substance in the tests here. I don't think we have to test the stringer functionality. 

Could we test that the right commands are sent though?

File: cmd/config/config.go (hunk 21 above)
Comment (jmorganca) [reply cmd/config/config.go]: I think we can get away with a simple ephemeral (optional) file/section that has the most recently used model for each integration in ~/.ollama/config.json or similar. Storing the models here isn't ideal because for integrations where we edit files we'll now have two lists of models (when one is better)

It can probably also remove a bunch of code here

File: cmd/config/config.go [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+
+	result := make([]integrationConfig, 0, len(configs))
+	for _, config := range configs {
+		result = append(result, *config)
+	}
+
+	return result, nil
+}
+
+var errEmptyAppName = errors.New("app name cannot be empty")
```
Comment (jmorganca) [inline cmd/config/config.go]: needed?

File: cmd/config/droid.go [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+	if err := os.MkdirAll(filepath.Dir(settingsPath), 0o755); err != nil {
+		return err
+	}
+
+	settings := make(map[string]any)
+	if data, err := os.ReadFile(settingsPath); err == nil {
+		_ = json.Unmarshal(data, &settings) // Ignore parse errors; treat missing/corrupt files as empty
+	}
+
+	customModels, _ := settings["customModels"].([]any)
```
Comment (jmorganca) [inline cmd/config/droid.go:86]: should we check OK here?

File: cmd/config/droid.go [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	settingsPath := filepath.Join(home, ".factory", "settings.json")
+	if err := os.MkdirAll(filepath.Dir(settingsPath), 0o755); err != nil {
+		return err
+	}
+
+	settings := make(map[string]any)
+	if data, err := os.ReadFile(settingsPath); err == nil {
+		_ = json.Unmarshal(data, &settings) // Ignore parse errors; treat missing/corrupt files as empty
```
Comment (jmorganca) [inline cmd/config/droid.go]: Do we want to error here? this seems dangerous to treat it as empty and the user might lose data

File: cmd/config/droid.go [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+
+	// Keep only non-Ollama models (we'll rebuild Ollama models fresh)
+	nonOllamaModels := slices.DeleteFunc(slices.Clone(customModels), isOllamaModelEntry)
+
+	// Build new Ollama model entries with sequential indices (0, 1, 2, ...)
+	var ollamaModels []any
+	var defaultModelID string
+	for i, model := range models {
+		modelID := fmt.Sprintf("custom:%s-[Ollama]-%d", model, i)
+		newEntry := map[string]any{
```
Comment (jmorganca) [inline cmd/config/droid.go]: Can we make this a struct vs a loose map[string]any? or is that too restrictive?

File: cmd/config/files.go [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
```
Comment (jmorganca) [inline cmd/config/files.go]: This is misleading since it doesn't lock the file in any way. Instead, just call this `writeWithBackup` or similar.

File: cmd/config/files_test.go [hunk 30]
```diff
@@ -0,0 +1,337 @@
...
+	// Original file should be preserved
+	current, _ := os.ReadFile(path)
+	if string(current) != string(originalContent) {
+		t.Errorf("original file was modified despite backup failure: got %s", string(current))
+	}
+}
+
+// TestAtomicWrite_PermissionDenied verifies clear error when target file has wrong permissions.
+// Common issue when config owned by root or wrong perms.
+func TestAtomicWrite_PermissionDenied(t *testing.T) {
```
Comment (jmorganca) [inline cmd/config/files_test.go]: Good tests here for edge cases. I'd see if we can think of any more that might happen.

File: cmd/config/integrations.go [hunk 31]
```diff
@@ -0,0 +1,350 @@
...
+	"github.com/spf13/cobra"
+)
+
+// Runners execute the launching of a model with the integration - claude, codex
+// Editors can edit config files (supports multi-model selection) - opencode, droid
+// They are composable interfaces where in some cases an editor is also a runner - opencode, droid
+// Runner can run an integration with a model.
+type Runner interface {
+	Run(model string) error
+	String() string
```
Comment (jmorganca) [inline cmd/config/integrations.go:27]: I don't think you need this line?

File: cmd/config/integrations.go [hunk 32]
```diff
@@ -0,0 +1,350 @@
...
+
+// Editor can edit config files (supports multi-model selection).
+type Editor interface {
+	Paths() []string
+	Edit(models []string) error
+	Models() []string
+}
+
+// integrations is the registry of available integrations.
+var integrations = map[string]Runner{
```
Comment (jmorganca) [inline cmd/config/integrations.go:41]: I know I suggested using `Runner` here but you really want this to be an integration vs a runner - since it forces all of these to be integrations that can be run. I think this is ok for now given they are all runnable, but thought for a the future

File: cmd/config/droid.go (hunk 27 above)
Comment (ParthSareen) [reply cmd/config/droid.go]: Hmmm this is on read so initially i was thinking okay to keep this way. This would happen if the file existed but was empty/not json/corrupted. I can error on it instead and point them to where to fix?

File: cmd/config/integrations.go [hunk 33]
```diff
@@ -0,0 +1,350 @@
...
+		}
+
+		var aErr api.AuthorizationError
+		if !errors.As(err, &aErr) || aErr.SigninURL == "" {
+			return nil, err
+		}
+
+		yes, err := confirmPrompt("Sign in to ollama.com?")
+		if err != nil || !yes {
+			return nil, fmt.Errorf("sign in required for cloud models")
```
Comment (jmorganca) [inline cmd/config/integrations.go]: "This model requires signing in"

File: cmd/config/integrations.go [hunk 34]
```diff
@@ -0,0 +1,350 @@
...
+		if err == nil && user != nil && user.Name != "" {
+			return selected, nil
+		}
+
+		var aErr api.AuthorizationError
+		if !errors.As(err, &aErr) || aErr.SigninURL == "" {
+			return nil, err
+		}
+
+		yes, err := confirmPrompt("Sign in to ollama.com?")
```
Comment (jmorganca) [inline cmd/config/integrations.go]:
```suggestion
		yes, err := confirmPrompt("Sign in?")
```

I'd remove ollama.com. However, should we have any other context here? e.g. "This model requires signing in, sign in now?"

File: cmd/config/integrations.go (hunk 31 above)
Comment (ParthSareen) [reply cmd/config/integrations.go:27]: `String()` just grabs the human readable name - can differentiate between the command vs. what should appear on the terminal. thought it was okay to keep

File: cmd/config/integrations.go [hunk 35]
```diff
@@ -0,0 +1,350 @@
...
+		if err != nil || !yes {
+			return nil, fmt.Errorf("sign in required for cloud models")
+		}
+
+		fmt.Fprintf(os.Stderr, "\nTo sign in, navigate to:\n    %s\n\n", aErr.SigninURL)
+
+		// Auto-open browser (best effort, fail silently)
+		switch runtime.GOOS {
```
Comment (jmorganca) [inline cmd/config/integrations.go:182]: Not important now but I'd add a TODO to extract this auth signin code sooner than later

File: cmd/config/integrations.go (hunk 34 above)
Comment (ParthSareen) [reply cmd/config/integrations.go]: Agree - let me think of something a bit cleaner

File: cmd/config/claude.go (hunk 23 above)
Comment (ParthSareen) [reply cmd/config/claude.go]: No i think we'll have to change the arg parsing

File: cmd/config/files.go [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+func readJSONFile(path string) (map[string]any, error) {
```
Comment (jmorganca) [inline cmd/config/files.go:12]: Instead of this, I'd open the file and pass it to the json unmarshaller

File: cmd/config/claude_test.go [hunk 37]
```diff
@@ -0,0 +1,19 @@
+package config
+
+import (
+	"testing"
+)
+
+func TestClaudeIntegration(t *testing.T) {
+	c := &Claude{}
+
+	t.Run("String", func(t *testing.T) {
```
Comment (jmorganca) [inline cmd/config/claude_test.go:11]: test that env vars get sent through

File: cmd/config/config.go [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+
+func integrationsPath() (string, error) {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(home, ".ollama", "config", "config.json"), nil
+}
+
+func loadIntegrationsFile() (*config, error) {
```
Comment (jmorganca) [inline cmd/config/config.go]: probably can call this `load`

File: cmd/config/config.go [hunk 39]
```diff
@@ -0,0 +1,115 @@
...
+	if err := json.Unmarshal(data, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse integrations file: %w, at: %s", err, path)
+	}
+	if cfg.Integrations == nil {
+		cfg.Integrations = make(map[string]*integration)
+	}
+	return &cfg, nil
+}
+
+func saveIntegrationsFile(cfg *config) error {
```
Comment (jmorganca) [inline cmd/config/config.go]: `save` ?

File: cmd/config/config.go [hunk 40]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	cfg.Integrations[strings.ToLower(appName)] = &integration{
+		Models: models,
+	}
+
+	return saveIntegrationsFile(cfg)
+}
+
+func loadIntegration(appName string) (*integration, error) {
```
Comment (jmorganca) [inline cmd/config/config.go:89]: This might be overkill vs load + access the data

Comment (ParthSareen) [reply cmd/config/config.go:89]: adding some base functions but leaving this in for now. would have to do a bit of extra work on each of the call sites (every editor integration)

**Comment by gabe-l-hart** [comment]
I've got a small follow-up to this that allows the launch configs to respect `envconfig.Host()` and point at non-default ollama server instances: https://github.com/ollama/ollama/pull/13937
